pkg/services/oauthserver/external_service.go

```diff
@@ -0,0 +1,161 @@
+	}
+	if len(c.PublicPem) > 0 {
+		c2.KeyResult = &KeyResult{PublicPem: string(c.PublicPem)}
+	}
+	return &c2
+}
+
+func (c *ExternalService) LogID() string {
+	return "{name: " + c.Name + ", clientID: " + c.ClientID + "}"
+}
+
+// GetID returns the client ID.
+func (c *ExternalService) GetID() string { return c.ClientID }
+
+// GetHashedSecret returns the hashed secret as it is stored in the store.
+func (c *ExternalService) GetHashedSecret() []byte {
+	// Hashed version is stored in the secret field
+	return []byte(c.Secret)
+}
+
+// GetRedirectURIs returns the client's allowed redirect URIs.
+func (c *ExternalService) GetRedirectURIs() []string {
+	return []string{c.RedirectURI}
+}
+
+// GetGrantTypes returns the client's allowed grant types.
+func (c *ExternalService) GetGrantTypes() fosite.Arguments {
+	return strings.Split(c.GrantTypes, ",")
+}
+
+// GetResponseTypes returns the client's allowed response types.
+// All allowed combinations of response types have to be listed, each combination having
+// response types of the combination separated by a space.
+func (c *ExternalService) GetResponseTypes() fosite.Arguments {
+	return fosite.Arguments{"code"}
+}
+
+// GetScopes returns the scopes this client is allowed to request on its own behalf.
+func (c *ExternalService) GetScopes() fosite.Arguments {
+	if c.Scopes != nil {
+		return c.Scopes
+	}
+
+	ret := []string{"profile", "email", "groups", "entitlements"}
+	if c.SignedInUser != nil && c.SignedInUser.Permissions != nil {
+		perms := c.SignedInUser.Permissions[TmpOrgID]
+		for action := range perms {
+			// Add all actions that the plugin is allowed to request
+			ret = append(ret, action)
+		}
+	}
+
+	c.Scopes = ret
+	return ret
+}
+
+// GetScopes returns the scopes this client is allowed to request on a specific user.
+func (c *ExternalService) GetScopesOnUser(ctx context.Context, accessControl ac.AccessControl, userID int64) []string {
+	ev := ac.EvalPermission(ac.ActionUsersImpersonate, ac.Scope("users", "id", strconv.FormatInt(userID, 10)))
+	hasAccess, errAccess := accessControl.Evaluate(ctx, c.SignedInUser, ev)
+	if errAccess != nil || !hasAccess {
+		return nil
+	}
+
+	if c.ImpersonateScopes != nil {
+		return c.ImpersonateScopes
+	}
+
+	ret := []string{}
+	if c.ImpersonatePermissions != nil {
+		perms := c.ImpersonatePermissions
+		for i := range perms {
+			if perms[i].Action == ac.ActionUsersRead && perms[i].Scope == ScopeGlobalUsersSelf {
+				ret = append(ret, "profile", "email", ac.ActionUsersRead)
+				continue
+			}
+			if perms[i].Action == ac.ActionUsersPermissionsRead && perms[i].Scope == ScopeUsersSelf {
+				ret = append(ret, "entitlements", ac.ActionUsersPermissionsRead)
+				continue
+			}
+			if perms[i].Action == ac.ActionTeamsRead && perms[i].Scope == ScopeTeamsSelf {
+				ret = append(ret, "groups", ac.ActionTeamsRead)
+				continue
+			}
+			// Add all actions that the plugin is allowed to request
+			ret = append(ret, perms[i].Action)
```

Comment:
There could be duplicates, but I don't think this is a big deal

Reply:
ack